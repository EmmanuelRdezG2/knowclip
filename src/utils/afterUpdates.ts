import { defer, Observable } from 'rxjs'
import { mergeAll } from 'rxjs/operators'

export const afterUpdates = <T>(callback: () => Promise<Observable<T>>) =>
  defer(async () => {
    await null
    return await callback()
  }).pipe(mergeAll())
