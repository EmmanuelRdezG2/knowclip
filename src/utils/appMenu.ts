import electron from 'electron'

export function setAppMenuProjectSubmenuPermissions(projectOpened: boolean) {
  const menu = electron.remote.Menu.getApplicationMenu()
  const submenu = menu && menu.getMenuItemById('File').submenu
  if (!submenu) return

  submenu.getMenuItemById('Save project').enabled = projectOpened
  submenu.getMenuItemById('Close project').enabled = projectOpened

  submenu.getMenuItemById('Open project').enabled = !projectOpened
}
