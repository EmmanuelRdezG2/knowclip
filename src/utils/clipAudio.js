import ffmpeg, { toTimestamp } from '../utils/ffmpeg'

const clipAudio = (sourceFilePath, startTime, endTime, outputFilename) => {
  return new Promise((res, rej) => {
    ffmpeg(sourceFilePath)
      // .audioCodec('copy') // later, do this and change hardcoded '.mp3' for audio-only input
      .seekInput(toTimestamp(startTime))
      .inputOptions('-to ' + toTimestamp(endTime))
      .outputOptions('-vn')
      .output(outputFilename)
      .on(
        'end',
        //listener must be a function, so to return the callback wrapping it inside a function
        function() {
          res(outputFilename)
        }
      )
      .on('error', err => {
        console.error(err)
        rej(err)
      })
      .run()
  })
}
export default clipAudio
