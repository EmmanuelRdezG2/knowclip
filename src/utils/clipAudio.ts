import ffmpeg, { toTimestamp } from '../utils/ffmpeg'
import { existsSync } from 'fs-extra'

import { Writable } from 'stream'

export const clipAudioStream = (
  sourceFilePath: string,
  startTime: number,
  endTime: number,
  outputFilename: string
): { done: Promise<string>; stream: Writable } => {
  const mp3 = ffmpeg(sourceFilePath)
    // .audioCodec('copy') // TODO: do this and change hardcoded '.mp3' for audio-only input
    .seekInput(toTimestamp(startTime))
    .inputOptions('-to ' + toTimestamp(endTime))
    .outputOptions('-vn')
    .outputFormat('mp3')
  return {
    done: new Promise((res, rej) => {
      mp3
        .on('end', () => {
          console.log('done!', outputFilename)
          res(outputFilename)
        })
        .on('error', err => {
          console.error(err)
          rej(err)
        })
    }),
    stream: mp3.pipe(),
  }
}

const clipAudio = async (
  sourceFilePath: string,
  startTime: number,
  endTime: number,
  outputFilename: string
): Promise<string> => {
  const result = await new Promise((res, rej) => {
    ffmpeg(sourceFilePath)
      // .audioCodec('copy') // TODO: do this and change hardcoded '.mp3' for audio-only input
      .seekInput(toTimestamp(startTime))
      .inputOptions('-to ' + toTimestamp(endTime))
      .outputOptions('-vn')
      .output(outputFilename)
      .on('end', () => {
        res(outputFilename)
      })
      .on('error', err => {
        console.error(err)
        rej(err)
      })
      .run()
  })

  if (!result || !existsSync(outputFilename)) {
    console.log({ outputFilename, sourceFilePath, startTime, endTime })
    throw new Error(`Problem clipping audio from ${sourceFilePath}`)
  }

  return outputFilename
}
export default clipAudio
