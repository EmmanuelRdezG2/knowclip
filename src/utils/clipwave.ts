export const CLIPWAVE_ID = 'waveform' as const
