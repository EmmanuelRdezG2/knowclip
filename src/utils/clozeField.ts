type RawClozeRange = {
  startIndex: number
  fullMatchText: string
  tagContents: string
}

export const unescapeClozeFields = <F extends FlashcardFields>(
  fieldsFromJson?: Partial<F>
): { fields: Partial<F>; cloze: ClozeDeletion[] } => {
  console.log('UNESCAPING!!')
  const fields = {} as F
  if (!fieldsFromJson) return { fields, cloze: [] }
  const cloze: ClozeDeletion[] = []
  for (const fn in fieldsFromJson) {
    const fieldName: keyof F = fn
    const value: string = fieldsFromJson[fieldName] as any
    if (fieldName === 'transcription') {
      const { text, clozeDeletions } = unescapeClozeField(value)
      fields[fieldName] = (text as unknown) as F[keyof F]
      cloze.push(...clozeDeletions)
    } else {
      fields[fieldName] = (value as unknown) as F[keyof F]
    }
  }

  return { fields, cloze }
}

const clozeTagsRegex = new RegExp(`<c(10|[1-9])>(.*?)</c>`, 'g')

export function unescapeClozeField(text: string) {
  const matches = [...text.matchAll(clozeTagsRegex)]

  let result = ''
  const rawRangeGroups = matches
    .reduce(
      (deletions, match) => {
        const [fullMatchText, clozeNumberRaw, tagContents] = match
        const index = typeof match.index === 'number' ? match.index : -1
        const clozeNumber = +clozeNumberRaw

        deletions[clozeNumber - 1] = deletions[clozeNumber - 1] || []
        deletions[clozeNumber - 1].push({
          startIndex: index,
          fullMatchText,
          tagContents,
        })
        return deletions
      },
      [] as RawClozeRange[][]
    )
    .filter(r => r && r.length)

  if (!rawRangeGroups.length) return { text, clozeDeletions: [] }

  const deletions: ClozeDeletion[] = rawRangeGroups.map(
    (rawRanges, rangeGroupIndex) => {
      const ranges: ClozeRange[] = []

      const firstRange = rawRanges[0]
      if (rangeGroupIndex === 0 && firstRange.startIndex > 0) {
        result += text.slice(0, firstRange.startIndex)
      }

      let i = 0
      for (const rawRange of rawRanges) {
        ranges.push({
          start: result.length,
          end: result.length + rawRange.tagContents.length,
        })
        result += rawRange.tagContents

        const nextRange = rawRanges[i + 1]
        const subsequentGapEnd = nextRange ? nextRange.startIndex : text.length
        result += text.slice(
          rawRange.startIndex + rawRange.fullMatchText.length,
          subsequentGapEnd
        )

        i++
      }

      return { ranges }
    }
  )

  return { text: result, clozeDeletions: deletions }
}
