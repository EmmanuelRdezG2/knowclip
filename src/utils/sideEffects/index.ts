import sideEffects from './module'
import spectronMocks from '../../test/spectronMocks'

const {
  mocked,
  mockFunctions: mockSideEffects,
  logMocks: logSideEffectsMocks,
} = spectronMocks('side-effect', sideEffects)

export { mockSideEffects, logSideEffectsMocks }

const { uuid, nowUtcTimestamp } =
  process.env.REACT_APP_CHROMEDRIVER || process.env.NODE_ENV === 'test'
    ? mocked
    : sideEffects

export { uuid, nowUtcTimestamp }
