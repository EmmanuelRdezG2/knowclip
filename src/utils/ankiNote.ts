import { join, basename } from 'path'
import { promises } from 'fs'
import clipAudio from '../utils/clipAudio'
import { getVideoStill } from '../utils/getVideoStill'
const { readFile } = promises

export type AnkiNoteMedia = {
  soundData: {
    data: () => Promise<Buffer>
    fileName: string
    filePath: string
  }
  imageData: {
    data: () => Promise<Buffer>
    fileName: string
    filePath: string
  } | null
}

export function getClozeFields(fields: string[], clozeDeletions: string) {
  return [clozeDeletions, ...fields.slice(1)]
}

export async function processNoteMedia(
  clipSpecs: ClipSpecs,
  directory: string
): AsyncResult<AnkiNoteMedia> {
  const {
    outputFilename,
    sourceFilePath,
    startTime,
    endTime,
    flashcardSpecs: { image },
  } = clipSpecs

  const clipOutputFilePath = join(directory, outputFilename)
  const clipAudioResult = await clipAudio(
    sourceFilePath,
    startTime,
    endTime,
    clipOutputFilePath
  )
  if (clipAudioResult.errors)
    return {
      errors: [
        `Could not make clip from ${sourceFilePath}`,
        ...clipAudioResult.errors,
      ],
    }
  const soundData = {
    data: () => readFile(clipOutputFilePath),
    fileName: outputFilename,
    filePath: clipOutputFilePath,
  }

  const imageResult = image
    ? await getVideoStill(image.id, sourceFilePath, image.seconds)
    : null
  if (imageResult && imageResult.errors)
    return {
      errors: [
        `Could not make still from ${sourceFilePath}`,
        ...imageResult.errors,
      ],
    }

  const imageData = imageResult
    ? {
        data: () => readFile(imageResult.value),
        fileName: basename(imageResult.value),
        filePath: imageResult.value,
      }
    : null

  return {
    value: {
      soundData,
      imageData,
    },
  }
}
