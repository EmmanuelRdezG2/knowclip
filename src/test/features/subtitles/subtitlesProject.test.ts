import { Application } from 'spectron'
import {
  startApp,
  stopApp,
  TestSetup,
  TMP_DIRECTORY,
  ASSETS_DIRECTORY,
  GENERATED_ASSETS_DIRECTORY,
} from '../../spectronApp'
import makeCardsFromSubtitles from './makeCardsFromSubtitles'
import saveAndCloseProject from './saveAndCloseProject'
import { join } from 'path'
import { mockSideEffects } from '../../../utils/sideEffects'

jest.setTimeout(60000)

const testId = 'subtitlesProject'

describe('make clips and cards from subtitles', () => {
  let context: { app: Application | null } = { app: null }
  let setup: TestSetup

  beforeAll(async () => {
    setup = await startApp(context, testId, persistedState)

    await mockSideEffects(setup.app, sideEffectsMocks)
  })

  test('automatically make clips and cards from subtitles', () =>
    makeCardsFromSubtitles(setup))
  test('saves and closes project', () => saveAndCloseProject(setup))

  afterAll(async () => {
    await stopApp(context)
  })
})

const sideEffectsMocks = {
  uuid: [
    '94367f0c-92ef-4542-a80a-6fbaf8e1ad76',
    'e78620d8-9ebd-45b8-b335-d3280a4f3ecc',
    '320265ad-8190-4729-8ec1-5fcf0d1b832e',
    '2f5ad340-c06b-4754-a468-56852ec9638c',
    '7d4e2afe-55d1-4a1a-8365-767aa83c83e5',
    '8a0d55ba-e456-4108-9fe2-911d7d25ffde',
    '193c54f4-49f0-442b-b866-320ba111d369',
    'c7e5886a-0899-4192-8a9c-4b30c304a717',
    'f4f17e82-af87-459e-af11-4f3d65f9d5d7',
    'f1def124-bbaa-418a-8836-34a8d8cf74da',
    '6c34f168-1e1d-4933-84ae-df589c68ec06',
    '24374cd7-bed9-4f30-aace-dbdf8fe31422',
    '8f0f15ae-fe46-4b73-ad7d-36c4d10359dd',
    '2a3ce4c2-4714-47f6-9e33-2e28ba56eb1f',
    '90bf45fc-00e0-41b7-b82e-fda29473761e',
  ],
  nowUtcTimestamp: ['2020-01-24T23:44:33Z', '2020-01-24T23:44:38Z'],
}

const persistedState: Partial<AppState> = {
  files: {
    ProjectFile: {
      'e36ca6de-4893-49f5-bca1-4410ace25d46': {
        id: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
        type: 'ProjectFile',
        lastSaved: '2020-01-22T13:43:37Z',
        lastOpened: '2020-01-24T15:09:03Z',
        name: 'Project with subtitles',
        mediaFileIds: [
          'fa462524-bfeb-4434-a506-07858cc97151',
          '535935a3-7d72-4238-87ab-1b7a413c1f71',
        ],
        error: null,
        noteType: 'Transliteration',
      },
    },
    MediaFile: {
      'fa462524-bfeb-4434-a506-07858cc97151': {
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
        type: 'MediaFile',
        parentId: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
        subtitles: [],
        flashcardFieldsToSubtitlesTracks: {},
        name: 'piggeldy_cat.mp4',
        durationSeconds: 58.24,
        format: 'mov,mp4,m4a,3gp,3g2,mj2',
        isVideo: true,
        subtitlesTracksStreamIndexes: [],
      },
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        type: 'MediaFile',
        parentId: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
        subtitles: [
          {
            type: 'EmbeddedSubtitlesTrack',
            id: '37bd5e91-89c5-491b-9308-533c7be7338a',
            streamIndex: 2,
          },
          {
            type: 'ExternalSubtitlesTrack',
            id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
          },
        ],
        flashcardFieldsToSubtitlesTracks: {
          transcription: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
          meaning: '37bd5e91-89c5-491b-9308-533c7be7338a',
        },
        name: 'polar_bear_cafe.mp4',
        durationSeconds: 87.062,
        format: 'mov,mp4,m4a,3gp,3g2,mj2',
        isVideo: true,
        subtitlesTracksStreamIndexes: [2],
      },
    },
    ExternalSubtitlesFile: {
      '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439': {
        type: 'ExternalSubtitlesFile',
        parentId: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
        name: 'pbc_jp.ass',
      },
    },
    VttConvertedSubtitlesFile: {
      '37bd5e91-89c5-491b-9308-533c7be7338a': {
        type: 'VttConvertedSubtitlesFile',
        parentId: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        id: '37bd5e91-89c5-491b-9308-533c7be7338a',
        streamIndex: 2,
        parentType: 'MediaFile',
      },
      '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439': {
        type: 'VttConvertedSubtitlesFile',
        id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
        parentId: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
        parentType: 'ExternalSubtitlesFile',
      },
    },
    WaveformPng: {
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        type: 'WaveformPng',
        parentId: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
      },
      'fa462524-bfeb-4434-a506-07858cc97151': {
        type: 'WaveformPng',
        parentId: 'fa462524-bfeb-4434-a506-07858cc97151',
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
      },
    },
    ConstantBitrateMp3: {},
  },
  fileAvailabilities: {
    ProjectFile: {
      'e36ca6de-4893-49f5-bca1-4410ace25d46': {
        filePath: join(TMP_DIRECTORY, 'project_with_subtitles.afca'),
        status: 'CURRENTLY_LOADED',
        id: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
      },
    },
    MediaFile: {
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        filePath: join(ASSETS_DIRECTORY, 'polar_bear_cafe.mp4'),
        status: 'CURRENTLY_LOADED',
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
      },
      'fa462524-bfeb-4434-a506-07858cc97151': {
        filePath: join(ASSETS_DIRECTORY, 'piggeldy_cat.mp4'),
        status: 'CURRENTLY_LOADED',
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
      },
    },
    ExternalSubtitlesFile: {},
    VttConvertedSubtitlesFile: {
      '37bd5e91-89c5-491b-9308-533c7be7338a': {
        filePath: join(
          GENERATED_ASSETS_DIRECTORY,
          testId,
          'cafbc25b9c6405df0dd0c77173539bcd.vtt'
        ),
        status: 'CURRENTLY_LOADED',
        id: '37bd5e91-89c5-491b-9308-533c7be7338a',
      },
    },
    WaveformPng: {
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        filePath: join(
          GENERATED_ASSETS_DIRECTORY,
          testId,
          '68a8d8c5862e344e7d032006fe1d2c0a.png'
        ),
        status: 'CURRENTLY_LOADED',
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
      },
      'fa462524-bfeb-4434-a506-07858cc97151': {
        filePath: join(
          GENERATED_ASSETS_DIRECTORY,
          testId,
          '992896afa3b7fd044dd8a60bdba4c730.png'
        ),
        status: 'CURRENTLY_LOADED',
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
      },
    },
    ConstantBitrateMp3: {},
  },
}
