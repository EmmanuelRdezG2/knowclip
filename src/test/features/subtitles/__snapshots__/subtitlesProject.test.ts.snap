// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`make clips and cards from subtitles saves and closes project 1`] = `
Object {
  "clips": Array [
    Object {
      "end": 888,
      "fileId": "fa462524-bfeb-4434-a506-07858cc97151",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "Du hast eben gesagt, murrte Piggeldy, eine Katze sieht aus wie ein Schwein, stimmt's?",
        },
        "id": "2f957585-b782-45fa-b0e7-ad6b0486d9ca",
        "tags": Array [
          "piggeldy_cat.mp4",
        ],
        "type": "Transliteration",
      },
      "id": "2f957585-b782-45fa-b0e7-ad6b0486d9ca",
      "start": 603,
    },
    Object {
      "end": 1545,
      "fileId": "fa462524-bfeb-4434-a506-07858cc97151",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "Sie hat nur fast alles, was ein Schwein auch hat. Aber sie spricht anders, sie sagt »Miau, miau, miau!«",
        },
        "id": "298d069b-1b30-420c-b3b4-2728569a6718",
        "tags": Array [
          "piggeldy_cat.mp4",
        ],
        "type": "Transliteration",
      },
      "id": "298d069b-1b30-420c-b3b4-2728569a6718",
      "start": 1074,
    },
    Object {
      "end": 2048,
      "fileId": "fa462524-bfeb-4434-a506-07858cc97151",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "Schön! seufzte Piggeldy verzückt.

Schön, wie du eben »Miau!« gemacht hast. Ein Glück, dass du keine Mäuse frisst.",
        },
        "id": "505afa2a-928d-4ce1-a741-8827c4e9a864",
        "tags": Array [
          "piggeldy_cat.mp4",
        ],
        "type": "Transliteration",
      },
      "id": "505afa2a-928d-4ce1-a741-8827c4e9a864",
      "start": 1571,
    },
    Object {
      "end": 626,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Chilling and eating bamboo 
grass is the best thing ever.",
          "notes": "",
          "pronunciation": "",
          "transcription": "笹を食べながらのんびりするのは最高だなぁ",
        },
        "id": "94367f0c-92ef-4542-a80a-6fbaf8e1ad76",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "94367f0c-92ef-4542-a80a-6fbaf8e1ad76",
      "start": 381.5,
    },
    Object {
      "end": 882.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Hey now. Look at you, being so lazy 
so early in the day.",
          "notes": "",
          "pronunciation": "",
          "transcription": "ちょっと  またこの子は昼間からゴロゴロして",
        },
        "id": "e78620d8-9ebd-45b8-b335-d3280a4f3ecc",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "e78620d8-9ebd-45b8-b335-d3280a4f3ecc",
      "start": 639,
    },
    Object {
      "end": 1063,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "You're in the way of my cleaning. Scoot.",
          "notes": "",
          "pronunciation": "",
          "transcription": "掃除の邪魔よ  どきなさい",
        },
        "id": "320265ad-8190-4729-8ec1-5fcf0d1b832e",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "320265ad-8190-4729-8ec1-5fcf0d1b832e",
      "start": 916.5,
    },
    Object {
      "end": 1136,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Okay.",
          "notes": "",
          "pronunciation": "",
          "transcription": "はいー",
        },
        "id": "2f5ad340-c06b-4754-a468-56852ec9638c",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "2f5ad340-c06b-4754-a468-56852ec9638c",
      "start": 1064,
    },
    Object {
      "end": 1372.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Seconds please, Mom.",
          "notes": "",
          "pronunciation": "",
          "transcription": "お母さん  おかわり",
        },
        "id": "7d4e2afe-55d1-4a1a-8365-767aa83c83e5",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "7d4e2afe-55d1-4a1a-8365-767aa83c83e5",
      "start": 1240,
    },
    Object {
      "end": 1609.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Exactly what do you have 
planned for your future?",
          "notes": "",
          "pronunciation": "",
          "transcription": "どうするの  あなた将来は",
        },
        "id": "8a0d55ba-e456-4108-9fe2-911d7d25ffde",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "8a0d55ba-e456-4108-9fe2-911d7d25ffde",
      "start": 1466,
    },
    Object {
      "end": 1961,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Jeez... Could we not talk about my future when I'm relaxing?",
          "notes": "",
          "pronunciation": "",
          "transcription": "もう  ゴロゴロしてるときに  将来の話なんかしないでよ",
        },
        "id": "193c54f4-49f0-442b-b866-320ba111d369",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "193c54f4-49f0-442b-b866-320ba111d369",
      "start": 1661.5,
    },
    Object {
      "end": 2131.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "If you have so much time to relax, you could go look for a part-time job.",
          "notes": "",
          "pronunciation": "",
          "transcription": "暇があるなら  ちょっとはバイトでも探しなさい",
        },
        "id": "c7e5886a-0899-4192-8a9c-4b30c304a717",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "c7e5886a-0899-4192-8a9c-4b30c304a717",
      "start": 1962.5,
    },
    Object {
      "end": 2464.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "I'm too busy relaxing, 
so I'll do it later.",
          "notes": "",
          "pronunciation": "",
          "transcription": "ええ  ゴロゴロするのに忙しいから  後にするよ",
        },
        "id": "f4f17e82-af87-459e-af11-4f3d65f9d5d7",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "f4f17e82-af87-459e-af11-4f3d65f9d5d7",
      "start": 2136.5,
    },
    Object {
      "end": 2755,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Don't try to suck me up!",
          "notes": "",
          "pronunciation": "",
          "transcription": "ああー  吸わないで",
        },
        "id": "f1def124-bbaa-418a-8836-34a8d8cf74da",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "f1def124-bbaa-418a-8836-34a8d8cf74da",
      "start": 2516.5,
    },
    Object {
      "end": 2912,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Jeez...",
          "notes": "",
          "pronunciation": "",
          "transcription": "もう",
        },
        "id": "6c34f168-1e1d-4933-84ae-df589c68ec06",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "6c34f168-1e1d-4933-84ae-df589c68ec06",
      "start": 2844,
    },
    Object {
      "end": 3233.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "I wonder why Mom can't understand how much time relaxing takes up.",
          "notes": "",
          "pronunciation": "",
          "transcription": "どうしてお母さんはゴロゴロする忙しさが分かんないんだろな",
        },
        "id": "24374cd7-bed9-4f30-aace-dbdf8fe31422",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "24374cd7-bed9-4f30-aace-dbdf8fe31422",
      "start": 2951,
    },
    Object {
      "end": 4099,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Excuse me, which of these will have jobs where I can get paid to do nothing?",
          "notes": "",
          "pronunciation": "",
          "transcription": "えーと  働かないでお金がもらえる仕事が載っているのはどれですか",
        },
        "id": "8f0f15ae-fe46-4b73-ad7d-36c4d10359dd",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "8f0f15ae-fe46-4b73-ad7d-36c4d10359dd",
      "start": 3764.5,
    },
    Object {
      "end": 4254.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "I don't think there is one.",
          "notes": "",
          "pronunciation": "",
          "transcription": "ないと思います",
        },
        "id": "2a3ce4c2-4714-47f6-9e33-2e28ba56eb1f",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "2a3ce4c2-4714-47f6-9e33-2e28ba56eb1f",
      "start": 4173,
    },
    Object {
      "end": 4336.5,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Really?",
          "notes": "",
          "pronunciation": "",
          "transcription": "え",
        },
        "id": "90bf45fc-00e0-41b7-b82e-fda29473761e",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "90bf45fc-00e0-41b7-b82e-fda29473761e",
      "start": 4268,
    },
  ],
  "id": "e36ca6de-4893-49f5-bca1-4410ace25d46",
  "lastOpened": "2020-01-24T23:44:33Z",
  "mediaFiles": Array [
    Object {
      "durationSeconds": 58.24,
      "flashcardFieldsToSubtitlesTracks": Object {},
      "format": "mov,mp4,m4a,3gp,3g2,mj2",
      "id": "fa462524-bfeb-4434-a506-07858cc97151",
      "isVideo": true,
      "name": "piggeldy_cat.mp4",
      "parentId": "e36ca6de-4893-49f5-bca1-4410ace25d46",
      "subtitles": Array [],
      "subtitlesTracksStreamIndexes": Array [],
      "type": "MediaFile",
    },
    Object {
      "durationSeconds": 87.062,
      "flashcardFieldsToSubtitlesTracks": Object {
        "meaning": "37bd5e91-89c5-491b-9308-533c7be7338a",
        "transcription": "2f7a98c0-1bf8-44cf-9b0b-4af64fb86439",
      },
      "format": "mov,mp4,m4a,3gp,3g2,mj2",
      "id": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "isVideo": true,
      "name": "polar_bear_cafe.mp4",
      "parentId": "e36ca6de-4893-49f5-bca1-4410ace25d46",
      "subtitles": Array [
        Object {
          "id": "37bd5e91-89c5-491b-9308-533c7be7338a",
          "streamIndex": 2,
          "type": "EmbeddedSubtitlesTrack",
        },
        Object {
          "id": "2f7a98c0-1bf8-44cf-9b0b-4af64fb86439",
          "type": "ExternalSubtitlesTrack",
        },
      ],
      "subtitlesTracksStreamIndexes": Array [
        2,
      ],
      "type": "MediaFile",
    },
  ],
  "name": "Project with subtitles",
  "noteType": "Transliteration",
  "subtitles": Array [
    Object {
      "id": "2f7a98c0-1bf8-44cf-9b0b-4af64fb86439",
      "name": "pbc_jp.ass",
      "parentId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "type": "ExternalSubtitlesFile",
    },
  ],
  "tags": Array [
    "piggeldy_cat.mp4",
    "pbc",
  ],
  "timestamp": "2020-01-24T23:44:38Z",
  "version": "4.1.0",
}
`;
